Essentials/src/com/earth2me/essentials/AsyncTeleport.java

```diff
@@ -0,0 +1,419 @@
+        }
+        return false;
+    }
+
+    private boolean cooldownApplies() {
+        boolean applies = true;
+        String globalBypassPerm = "essentials.teleport.cooldown.bypass";
+        switch (tpType) {
+            case NORMAL:
+                applies = !teleportOwner.isAuthorized(globalBypassPerm);
+                break;
+            case BACK:
+                applies = !(teleportOwner.isAuthorized(globalBypassPerm) &&
+                    teleportOwner.isAuthorized("essentials.teleport.cooldown.bypass.back"));
+                break;
+            case TPA:
+                applies = !(teleportOwner.isAuthorized(globalBypassPerm) &&
+                    teleportOwner.isAuthorized("essentials.teleport.cooldown.bypass.tpa"));
+                break;
+        }
+        return applies;
+    }
+
+    private void warnUser(final IUser user, final double delay) {
+        Calendar c = new GregorianCalendar();
+        c.add(Calendar.SECOND, (int) delay);
+        c.add(Calendar.MILLISECOND, (int) ((delay * 1000.0) % 1000.0));
+        user.sendMessage(tl("dontMoveMessage", DateUtil.formatDateDiff(c.getTimeInMillis())));
+    }
+
+
+    @Override
+    public void now(Location loc, boolean cooldown, TeleportCause cause, CompletableFuture<Boolean> future) {
+        if (cooldown && cooldown(false, future)) {
+            return;
+        }
+        final ITarget target = new LocationTarget(loc);
+        nowAsync(teleportOwner, target, cause, future);
+    }
+
+    @Override
+    public void now(Player entity, boolean cooldown, TeleportCause cause, CompletableFuture<Boolean> future) {
+        if (cooldown && cooldown(false, future)) {
+            future.complete(false);
+            return;
+        }
+        final ITarget target = new PlayerTarget(entity);
+        nowAsync(teleportOwner, target, cause, future);
+        future.thenAccept(success -> {
+            if (success) {
+                teleportOwner.sendMessage(tl("teleporting", target.getLocation().getWorld().getName(), target.getLocation().getBlockX(), target.getLocation().getBlockY(), target.getLocation().getBlockZ()));
+            }
+        });
+    }
+
+    protected void nowAsync(IUser teleportee, ITarget target, TeleportCause cause, CompletableFuture<Boolean> future) {
+        cancel(false);
+
+        UserTeleportEvent event = new UserTeleportEvent(teleportee, cause, target.getLocation());
+        Bukkit.getServer().getPluginManager().callEvent(event);
+        if (event.isCancelled()) {
+            return;
+        }
+        teleportee.setLastLocation();
+
+        if (!teleportee.getBase().isEmpty()) {
+            if (!ess.getSettings().isTeleportPassengerDismount()) {
+                future.completeExceptionally(new Exception(tl("passengerTeleportFail")));
+                return;
+            }
+            CompletableFuture<Object> dismountFuture = new CompletableFuture<>();
+            Bukkit.getScheduler().runTask(ess, () -> {
```

Comment:
This logic is called on main, thus this will deadlock; need to check the current thread before doing this

Reply:
Something like this?
```java
            if (Bukkit.isPrimaryThread()) {
                teleportee.getBase().eject();
            } else {
                CompletableFuture<Object> dismountFuture = new CompletableFuture<>();
                Bukkit.getScheduler().runTask(ess, () -> {
                    teleportee.getBase().eject();
                    dismountFuture.complete(new Object());
                });
                try {
                    dismountFuture.get(); //EntityDismountEvent requires sync context we also want to wait for it to finish
                } catch (InterruptedException | ExecutionException e) {
                    future.completeExceptionally(e);
                    return;
                }
            }
```